src/player/systems/abilities.rs

```diff
@@ -0,0 +1,237 @@
+    for event in ability_event_reader.read() {
+        for (
+            action_state,
+            mut player_ext_impulse,
+            mut player_movement,
+            mut player_incoming_damage,
+            player_id,
+            children,
+        ) in player_query.iter_mut()
+        {
+            for child in children.iter() {
+                if let Ok((ability_id, mut charge_ability)) = ability_query.get_mut(*child) {
+                    if event.player_id == *player_id && event.ability_slot_id == *ability_id {
+                        // check all movement inputs to see if the player wants to charge in a specific direction
+                        let up = action_state.pressed(&PlayerAction::MoveUp);
+                        let down = action_state.pressed(&PlayerAction::MoveDown);
+                        let left = action_state.pressed(&PlayerAction::MoveLeft);
+                        let right = action_state.pressed(&PlayerAction::MoveRight);
+
+                        // try to create a normalized Vec2 using the inputs
+                        if let Some(vec2_normal) = Vec2::new(
+                            (-(left as i8) + right as i8) as f32,
+                            (-(down as i8) + up as i8) as f32,
+                        )
+                        .try_normalize()
+                        {
+                            // multiply the normalized vector by the charge ability's impulse
+                            player_ext_impulse.impulse = charge_ability.impulse * vec2_normal;
+                        } else {
+                            // if a normalized vector could not be created apply the impulse in the +y direction
+                            player_ext_impulse.impulse = Vec2::new(0.0, charge_ability.impulse);
+                        }
+
+                        // disable movement and apply damage reduction
+                        player_movement.movement_enabled = false;
+                        player_incoming_damage.multiplier -=
+                            charge_ability.incoming_damage_multiplier;
+
+                        // begin the action timer for the ability
+                        charge_ability.action_timer.reset();
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Updates the charge ability (`ChargeAbilityComponent`)
+/// Ticks the action timer, when completed enables movment and resets the incoming damage multiplier
+pub fn update_charge_ability_system(
+    mut player_query: Query<(
+        &mut Velocity,
+        &mut PlayerMovementComponent,
+        &mut PlayerIncomingDamageComponent,
+        &Children,
+    )>,
+    mut ability_query: Query<&mut ChargeAbilityComponent>,
+    time: Res<Time>,
+) {
+    for (mut player_velocity, mut player_movement, mut player_incoming_damage, children) in
+        player_query.iter_mut()
+    {
+        for child in children.iter() {
+            if let Ok(mut charge_ability) = ability_query.get_mut(*child) {
```

Comment:
I made the decision to make abilities children of the player entity for a couple reasons.

1. Allows us to associate multiple components with a single player (like AbilityCooldownComponent)

2.  Allows easy swapping of abilities. Despawn old ability then spawn new ability as child in same ability slot

Also this is for players which will likely only go up to 4, if players each have 4 abilities this will only be 16 loops worst case.